File: samples/adapters/db_client.js
/**
 * DBClient implementation using PostgreSQL
 *
 * Implements the DBClient Required Port.
 */

const { Pool } = require('pg');

class DBAdapter {
  constructor(config = null) {
    this.pools = new Map();

    if (config) {
      this.createPool('default', config);
    }
  }

  /**
   * Create a connection pool
   * @param {string} name - Pool name
   * @param {Object} config - Connection configuration (complete config object from State)
   *
   * Expected config structure:
   * {
   *   host: string,
   *   port: number,
   *   database: string,
   *   username: string,
   *   password: string,
   *   driver: string,      // e.g., 'mysql', 'postgres'
   *   charset: string,     // e.g., 'utf8mb4'
   *   collation: string    // e.g., 'utf8mb4_unicode_ci'
   * }
   */
  createPool(name, config) {
    const poolConfig = {
      host: config.host || process.env.DB_HOST,
      port: config.port || process.env.DB_PORT || 5432,
      database: config.database || process.env.DB_DATABASE,
      user: config.username || process.env.DB_USERNAME,
      password: config.password || process.env.DB_PASSWORD,
    };

    // Add charset/collation if provided (for MySQL/MariaDB compatibility)
    if (config.charset) {
      poolConfig.charset = config.charset;
    }

    const pool = new Pool(poolConfig);
    this.pools.set(name, pool);
  }

  /**
   * Get pool by name
   * @param {string} name - Pool name
   * @returns {Pool} PostgreSQL pool
   */
  getPool(name = 'default') {
    if (!this.pools.has(name)) {
      throw new Error(`Pool '${name}' not found`);
    }
    return this.pools.get(name);
  }

  /**
   * Fetch records from database
   * @param {Object} config - Complete connection config object from State (not string!)
   * @param {string} table - Table name
   * @param {Array<string>} columns - Column names to SELECT (e.g., ['db_host', 'db_port'])
   * @param {string|null} whereClause - WHERE clause (e.g., "id=123")
   * @returns {Promise<Array>} Array of records (0 or more rows)
   *
   * NOTE: This implementation expects config to be an Object.
   * If you receive a string (connection name), you should maintain
   * your own connection map instead of calling State.
   *
   * SQL generation example:
   * fetch(config, 'tenants', ['db_host', 'db_port'], 'id=1')
   * → SELECT db_host, db_port FROM tenants WHERE id=1
   */
  async fetch(config, table, columns, whereClause = null) {
    const poolName = this.getPoolName(config);
    if (!this.pools.has(poolName)) {
      this.createPool(poolName, config);
    }

    const pool = this.getPool(poolName);

    // Build SELECT clause
    const columnList = columns.length > 0 ? columns.join(', ') : '*';
    let query = `SELECT ${columnList} FROM ${table}`;

    if (whereClause) {
      query += ` WHERE ${whereClause}`;
    }

    const result = await pool.query(query);
    return result.rows;
  }

  /**
   * Generate pool name from config
   * @param {Object} config - Connection config
   * @returns {string} Pool name
   */
  getPoolName(config) {
    return `${config.host}_${config.database}`;
  }

  /**
   * Close all pools
   */
  async closeAll() {
    for (const [name, pool] of this.pools) {
      await pool.end();
      console.log(`Pool '${name}' closed`);
    }
    this.pools.clear();
  }
}

module.exports = DBAdapter;

File: samples/adapters/db_client.rs
/// DBClient implementation using PostgreSQL
///
/// Implements the DBClient Required Port.

use state_engine::ports::required::DBClient;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Mutex;

pub struct DBAdapter {
    // Use Mutex for interior mutability with Send + Sync
    pool: Mutex<Option<tokio_postgres::Client>>,
}

impl DBAdapter {
    pub fn new() -> Self {
        Self {
            pool: Mutex::new(None),
        }
    }

    /// Create database connection from config
    async fn connect_from_config(config: &Value) -> Result<tokio_postgres::Client, String> {
        let config_obj = config.as_object()
            .ok_or("connection must be an object")?;

        let host = config_obj.get("host")
            .and_then(|v| v.as_str())
            .ok_or("Missing host in config")?;
        let port = config_obj.get("port")
            .and_then(|v| v.as_u64())
            .unwrap_or(5432) as u16;
        let database = config_obj.get("database")
            .and_then(|v| v.as_str())
            .ok_or("Missing database in config")?;
        let username = config_obj.get("username")
            .and_then(|v| v.as_str())
            .ok_or("Missing username in config")?;
        let password = config_obj.get("password")
            .and_then(|v| v.as_str())
            .ok_or("Missing password in config")?;

        let conn_str = format!(
            "host={} port={} dbname={} user={} password={}",
            host, port, database, username, password
        );

        let (client, connection) = tokio_postgres::connect(&conn_str, tokio_postgres::NoTls)
            .await
            .map_err(|e| format!("Failed to connect to database: {}", e))?;

        // Spawn connection handler
        tokio::spawn(async move {
            if let Err(e) = connection.await {
                eprintln!("Connection error: {}", e);
            }
        });

        Ok(client)
    }
}

impl DBClient for DBAdapter {
    fn fetch(
        &self,
        connection: &Value,
        table: &str,
        columns: &[&str],
        where_clause: Option<&str>,
    ) -> Option<Vec<HashMap<String, Value>>> {
        // Create runtime for async operation
        let runtime = tokio::runtime::Runtime::new().ok()?;

        runtime.block_on(async {
            // Ensure we have a connection
            let mut pool_lock = self.pool.lock().unwrap();
            if pool_lock.is_none() {
                let client = Self::connect_from_config(connection).await.ok()?;
                *pool_lock = Some(client);
            }

            let client = pool_lock.as_ref()?;

            // Build SELECT clause
            let column_list = if columns.is_empty() {
                "*".to_string()
            } else {
                columns.join(", ")
            };

            let query = if let Some(wc) = where_clause {
                format!("SELECT {} FROM {} WHERE {}", column_list, table, wc)
            } else {
                format!("SELECT {} FROM {}", column_list, table)
            };

            let rows = client.query(&query, &[]).await.ok()?;

            // Convert rows to Vec<HashMap<String, Value>>
            let mut results = Vec::new();
            for row in rows {
                let mut map = HashMap::new();
                for (idx, column) in row.columns().iter().enumerate() {
                    let value: Value = match column.type_() {
                        &tokio_postgres::types::Type::INT4 => {
                            row.try_get::<_, i32>(idx)
                                .map(|v| serde_json::json!(v))
                                .unwrap_or(Value::Null)
                        }
                        &tokio_postgres::types::Type::INT8 => {
                            row.try_get::<_, i64>(idx)
                                .map(|v| serde_json::json!(v))
                                .unwrap_or(Value::Null)
                        }
                        &tokio_postgres::types::Type::TEXT | &tokio_postgres::types::Type::VARCHAR => {
                            row.try_get::<_, String>(idx)
                                .map(|v| serde_json::json!(v))
                                .unwrap_or(Value::Null)
                        }
                        &tokio_postgres::types::Type::BOOL => {
                            row.try_get::<_, bool>(idx)
                                .map(|v| serde_json::json!(v))
                                .unwrap_or(Value::Null)
                        }
                        _ => Value::Null,
                    };
                    map.insert(column.name().to_string(), value);
                }
                results.push(map);
            }

            Some(results)
        })
    }
}
